lib/spack/spack/version.py

```diff
@@ -347,9 +347,12 @@ def __getitem__(self, idx):
                 string_arg.append(str(token))
                 string_arg.append(str(sep))
 
-            string_arg.pop()  # We don't need the last separator
-            string_arg = ''.join(string_arg)
-            return cls(string_arg)
+            if string_arg:
+                string_arg.pop()  # We don't need the last separator
+                string_arg = ''.join(string_arg)
+                return cls(string_arg)
+            else:
+                return None
```

Comment:
Can we return something that behaves like an empty list, rather than `None`? This would mimic the built-in `list` behavior:
```python
Python 3.8.10 (default, Jun  2 2021, 10:49:15) 
[GCC 9.4.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> l = [1, 2, 3]
>>> l[1:]
[2, 3]
>>> l[10:]
[]
```

Reply:
@tldahlgren I'm unsure what is the intended change of behavior for the Version class here. If I didn't miss something, this PR seems to raise a `ValueError` instead of an `IndexError` under the same condition as before. 

The question (question, not request) I have is if this class should behave like a list since it implements a `__getitem__` that accepts slices. Built-in lists behave like this with slices that go out of range:
```python
>>> l = [1,2]
>>> l[:3]
[1, 2]
>>> l[3:]
[]
```
This class on `develop` has the following results:
```
Spack version 0.16.3
Python 3.8.10, Linux x86_64
>>> import spack.version
>>> v = spack.version.Version('7.1')
>>> v[:2]
Version('7.1')
>>> v[:3]
Version('7.1')
>>> v[3:]
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/version.py", line 350, in __getitem__
    string_arg.pop()  # We don't need the last separator
IndexError: pop from empty list
>>> 
```
while on this branch it behaves like:
```
Spack version 0.16.3
Python 3.8.10, Linux x86_64
>>> import spack.version
>>> v = spack.version.Version('7.1')
>>> v[:3]
Version('7.1')
>>> v[3:]
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/version.py", line 354, in __getitem__
    return cls(string_arg)
  File "/home/culpo/PycharmProjects/spack/lib/spack/spack/version.py", line 181, in __init__
    raise ValueError("Bad characters in version string: %s" % string)
ValueError: Bad characters in version string: []
```
so it seems it just raises a different exception type. Would it make sense to add the "empty" version? Otherwise, what is the purpose of raising a different exception type? Pinging @haampie since he modified recently the Version class, so he might have opinions on this.

Reply:
I'll have to revisit.  I didn't re-test after the merge conflict post-requested change.